### PR TITLE
[Quant Tool] Introduce get_qdq_config() helper to get QDQ configurations

### PR DESCRIPTION
**Pull request by adrianlizarraga**
### Description
Introduces the `get_qdq_config()` function to get a quantization configuration for a full integer QDQ model. This function provides an easier way of specifying commonly used options and sets convenient defaults. Specifically:

- Instead of requiring the user to pass a dictionary of `extra_options`, the new interface adds function parameters for common settings:
  - All calibrator settings
  - Whether activations/weights are symmetric
  - Whether to keep or fuse relu/clip into Q
  - Minimum real range for quantization
  - Dictionary of tensor quantization overrides.
- Automatically scans the input floating-point model and fills out the operator types to quantize. Otherwise, only a limited number of operator types would be quantized by default.
- Detects if the input model uses external data. If so, ensures that the generated QDQ model also uses external data.
- Detects if the model will use newly introduced quantization types (int4/int16) with an older opset. If so, forces the use of the `com.microsoft` domain for Q/DQ ops, which support all types.
- Automatically enables the "extra option" called `ForceQuantizeNoInputCheck` to ensure data movement operators (e.g., Transpose) are always quantized.
- User can pass a function to indicate which nodes to exclude from quantization.
- The user can still pass their own `extra_options` to override any of the above if necessary.
 
```python
from onnxruntime.quantization import get_int_qdq_config, quantize # , ...

# Get QDQ configuration
qdq_config = get_int_qdq_config(
    float_model,
    data_reader,
    calibrate_method=CalibrationMethod.Percentile,
    calibrate_args={"percentile": 99.98},  # Converted to extra_options
    activation_type=QuantType.QUInt8,
    weight_type=QuantType.QInt8,
    per_channel=True,
    nodes_to_exclude=["Mul"], # Could also be a function. Ex: `lambda model, node: node.op_type == "Softmax"`

    # Other options converted to extra_options:
    min_real_range=0.0001,
    keep_removable_activations=True,
    activation_symmetric=True,
    weight_symmetric=True,
)

# Quantize model
quantize(float_model_path, qdq_model_path, qdq_config)
```
### Motivation and Context
Need a version of `get_qnn_qdq_config()` that is not EP-specific.

### CODE REVIEW

File: onnxruntime/test/python/quantization/test_get_int_qdq_config.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+
+
+class TestGetIntQDQConfig(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls._tmp_model_dir = tempfile.TemporaryDirectory(prefix="ort.int_qdq_config_")
+
+        # Note: swap with the commented line if you want to see the models in local test dir.
+        # cls._tmp_dir_path = cls._tmp_model_dir.name
+        cls._tmp_dir_path = "."
```
Comment (adrianlizarraga) [inline onnxruntime/test/python/quantization/test_get_int_qdq_config.py:27]:
```suggestion
        cls._tmp_dir_path = cls._tmp_model_dir.name
        # cls._tmp_dir_path = "."
```

File: onnxruntime/python/tools/quantization/quantize.py [hunk 2]
```diff
@@ -213,6 +217,164 @@ def __init__(
         self.extra_options = extra_options or {}
 
 
+def get_int_qdq_config(
```
Comment (fajin-corp) [inline onnxruntime/python/tools/quantization/quantize.py:220]: >get_int_qdq_config [](http://example.com/codeflow?start=4&length=18)

is there non-int qdq config?

File: onnxruntime/python/tools/quantization/quantize.py [hunk 3]
```diff
@@ -213,6 +217,164 @@ def __init__(
...
+        extra_options: Additional options specified as string key/value pairs. Refer to the documentation for
+            `quantize_static` for valid keys and values.
+
+    Returns:
+        A StaticQuantConfig object
+    """
+    q16_types = {QuantType.QInt16, QuantType.QUInt16}
+    q4_types = {QuantType.QInt4, QuantType.QUInt4}
+    op_types_to_exclude = {"Cast", "DequantizeLinear", "QuantizeLinear"}
+    model_size_threshold = 2147483648  # Quant model should use external data if >= 2GB
```
Comment (fajin-corp) [inline onnxruntime/python/tools/quantization/quantize.py:297]: >2147483648 [](http://example.com/codeflow?start=27&length=10)

nit: make it a constant

**Review by fajin-corp — APPROVED** [review]
:shipit:

File: onnxruntime/python/tools/quantization/quantize.py (hunk 2 above)
Comment (adrianlizarraga) [reply onnxruntime/python/tools/quantization/quantize.py:220]: Theres float8 quantization. This helper function was made with int only quantization in mind.

File: onnxruntime/python/tools/quantization/quantize.py (hunk 3 above)
Comment (adrianlizarraga) [reply onnxruntime/python/tools/quantization/quantize.py]: Made it a constant accessible from quant_utils.py

**Review by fajin-corp — APPROVED** [review]

**Comment by snnn** [comment]
This PR has been cherry-picked into the `rel-1.20.1` branch in PR #22785. Removing the `release:1.20.1` label.

**Pull request merged**